builder/core/project.py

```diff
@@ -410,9 +427,32 @@ def pre_build(self, env):
         build_deps = []
         for d in deps:
             dep_steps = _build_project(d, env)
+
             if dep_steps:
                 build_deps += [Script(dep_steps,
                                       name='build {}'.format(d.name))]
+
+            # allow root project to register additional pre/post build steps that run before or after the dependency is built
```

Comment:
So this will work, but I kind of wonder if we shouldn't just take the whole config tree and apply it, by ensuring that the upstream and downstream packages are cached with their contents before being resolved (which will cause a merge when the actual builder.json is loaded)